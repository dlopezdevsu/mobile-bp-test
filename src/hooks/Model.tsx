class model_class {

  dispatch: any = null
  get_dispatch = () => this.dispatch == null ? false : this.dispatch
  set_dispatch = (n_dispatch: any) => { this.dispatch = n_dispatch }

  setData = (label: string, payload: any) => {
    this?.dispatch({ type: 'SET_STORE', label, payload })
  }

  resetData = () => {
    this?.dispatch({ type: 'RESET_DATA' })
  }
}

export default new model_class();