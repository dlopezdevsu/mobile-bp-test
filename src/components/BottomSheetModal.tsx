import * as React from 'react';
import { View, StyleSheet, Text, Modal, Image, Pressable } from 'react-native';

import * as Colors from 'src/utils/colors';
import { InitialStateProps, Store } from 'src/hooks/main_store';
import { useContext } from 'react';
import Model from 'src/hooks/Model';
import Button from 'src/components/Button';
import Api from 'src/utils/api';

const BottomSheetModal = () => {
  const { state }: { state: InitialStateProps } = useContext(Store)

  const onConfirm = async () => {
    try {
      const response = (await (await Api.deleteProduct(state.selected?.id || '')).json())

      if (response?.message === "Product removed successfully") {
        state?.callbackBottomSheet?.()
      } else {
        throw new Error(response?.message)
      }
    } catch (error: any) {
      Model.setData('alert', error?.message || error)
    } finally {
      Model.setData('openBottomSheet', false)
    }
  }

  return (
    <>
      {state.openBottomSheet && <View style={styles.backdrop} />}

      <Modal
        animationType="slide"
        transparent={true}
        visible={state.openBottomSheet}
        onRequestClose={() => Model.setData('openBottomSheet', false)}>
        <View style={styles.container}>
          <View style={styles.header}>
            <Pressable style={({ pressed }) => [styles.containerIcon, { opacity: pressed ? 0.5 : 1 }]} onPress={() => Model.setData('openBottomSheet', false)} >

              <Image style={styles.icon} source={require('src/assets/icons/close.png')} />
            </Pressable>
          </View>
          <View style={styles.containerMessage}>
            <Text style={styles.message}>¿Estás seguro de eliminar el producto {state.selected?.name}?</Text>
          </View>
          <View style={styles.containerButtons}>
            <Button label='Confirmar' style={styles.confirmButton} onPress={onConfirm} variant='submit' />
            <Button label='Cancelar' style={styles.cancelButton} onPress={() => Model.setData('openBottomSheet', false)} />
          </View>
        </View>
      </Modal>
    </>
  );
};

const styles = StyleSheet.create({
  backdrop: {
    position: 'absolute',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
    width: '100%',
    height: '100%',
  },
  container: {
    flex: 1,
    justifyContent: 'flex-end'
  },
  header: {
    backgroundColor: Colors.white,
    borderTopLeftRadius: 12,
    borderTopRightRadius: 12,
    padding: 6,
    alignItems: 'flex-end'
  },
  containerIcon: {
    padding: 12
  },
  icon: {
    width: 18,
    height: 18,
    resizeMode: 'contain'
  },
  containerMessage: {
    backgroundColor: Colors.white,
    borderTopColor: Colors.gray,
    borderTopWidth: 1,
    borderBottomColor: Colors.gray,
    borderBottomWidth: 1,
    padding: 30
  },
  message: {
    textAlign: 'center',
    fontSize: 18
  },
  containerButtons: {
    backgroundColor: Colors.white,
    padding: 12
  },
  confirmButton: {
    marginVertical: 12,
  },
  cancelButton: {
    marginBottom: 18,
  },
});

export default BottomSheetModal;