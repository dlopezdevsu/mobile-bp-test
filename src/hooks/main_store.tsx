import React, { ReactElement } from "react";

import { ItemProductListProps } from 'src/components/ItemProductList'

export const Store = React.createContext<any>(null);

export interface InitialStateProps {
  loading: boolean;
  alert: string;
  products: ItemProductListProps[];
  selected: ItemProductListProps | null;
  editMode: boolean;
  openBottomSheet: boolean;
  callbackBottomSheet: () => void;
}

interface ActionProps {
  label: string;
  type: string;
  payload: any;
}

const initialState = {
  loading: true,
  alert: '',
  products: [],
  selected: null,
  editMode: false,
  openBottomSheet: false,
  callbackBottomSheet: () => null,
};

function reducer(state: InitialStateProps, action: ActionProps) {
  switch (action.type) {
    case "SET_STORE":
      return { ...state, [action.label]: action.payload };
    case "RESET_DATA":
      return { ...initialState };
    default:
      return state;
  }
}

export function StoreProvider({ children }: { children: ReactElement }) {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  const value = { state, dispatch };

  return <Store.Provider value={value}>{children}</Store.Provider>;
}
