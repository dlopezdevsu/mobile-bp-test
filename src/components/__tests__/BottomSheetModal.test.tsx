/**
 * @format
 */

import 'react-native';
import React from "react";
import BottomSheetModal from "src/components/BottomSheetModal";
import renderer from "react-test-renderer";
import { StoreProvider } from 'src/hooks/main_store'

import { it, expect } from '@jest/globals';

it(`render BottomSheetModal`, () => {
  const tree = renderer.create(<StoreProvider>
    <BottomSheetModal />
  </StoreProvider>);
  expect(tree).toMatchSnapshot();
});
