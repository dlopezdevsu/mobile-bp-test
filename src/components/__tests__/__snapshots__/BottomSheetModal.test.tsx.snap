// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render BottomSheetModal 1`] = `null`;
